src/keys.rs

```diff
@@ -1,70 +1,207 @@
 /// ability to spend funds (such as a view-only wallet).
-///
-/// TODO: Should we just define the FVK to include extended stuff like the diversifier key?
 #[derive(Debug)]
 pub struct FullViewingKey {
     ak: AuthorizingKey,
     nk: NullifierDerivingKey,
-    rivk: (),
+    rivk: CommitIvkRandomness,
 }
 
 impl From<&SpendingKey> for FullViewingKey {
-    fn from(_: &SpendingKey) -> Self {
-        todo!()
+    fn from(sk: &SpendingKey) -> Self {
+        FullViewingKey {
+            ak: (&SpendAuthorizingKey::from(sk)).into(),
+            nk: sk.into(),
+            rivk: sk.into(),
+        }
     }
 }
 
 impl FullViewingKey {
+    /// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
+    ///
+    /// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+    fn derive_dk_ovk(&self) -> (DiversifierKey, OutgoingViewingKey) {
+        let k = self.rivk.0.to_bytes();
+        let b = [(&self.ak.0).into(), self.nk.0.to_bytes()];
+        let r = prf_expand_vec(&k, &[&[0x82], &b[0][..], &b[1][..]]);
+        (
+            DiversifierKey(r.as_bytes()[..32].try_into().unwrap()),
+            OutgoingViewingKey(r.as_bytes()[32..].try_into().unwrap()),
+        )
+    }
+
+    /// Returns the default payment address for this key.
+    pub fn default_address(&self) -> Address {
+        self.address(DiversifierKey::from(self).default_diversifier())
+    }
+
     /// Returns the payment address for this key corresponding to the given diversifier.
     pub fn address(&self, d: Diversifier) -> Address {
         IncomingViewingKey::from(self).address(d)
     }
 }
 
+/// A key that provides the capability to derive a sequence of diversifiers.
+#[derive(Debug)]
+pub struct DiversifierKey([u8; 32]);
+
+impl From<&FullViewingKey> for DiversifierKey {
+    fn from(fvk: &FullViewingKey) -> Self {
+        fvk.derive_dk_ovk().0
+    }
+}
+
+/// The index for a particular diversifier.
+#[derive(Clone, Copy, Debug)]
+pub struct DiversifierIndex([u8; 11]);
+
+macro_rules! di_from {
+    ($n:ident) => {
+        impl From<$n> for DiversifierIndex {
+            fn from(j: $n) -> Self {
+                let mut j_bytes = [0; 11];
+                j_bytes[..mem::size_of::<$n>()].copy_from_slice(&j.to_le_bytes());
+                DiversifierIndex(j_bytes)
+            }
+        }
+    };
+}
+di_from!(u8);
+di_from!(u16);
+di_from!(u32);
+di_from!(u64);
+di_from!(usize);
+
+impl DiversifierKey {
+    /// Returns the diversifier at index 0.
+    pub fn default_diversifier(&self) -> Diversifier {
+        self.get(0u8)
+    }
+
+    /// Returns the diversifier at the given index.
+    pub fn get(&self, j: impl Into<DiversifierIndex>) -> Diversifier {
+        let ff = FF1::<Aes256>::new(&self.0, 2).expect("valid radix");
+        let enc = ff
+            .encrypt(&[], &BinaryNumeralString::from_bytes_le(&j.into().0[..]))
+            .unwrap();
```

Comment:
Any chance this unwrap will panic?

Reply:
No; the only way this API errors is if the radix doesn't match the numeral string, and we initialize both the radix and the numeral string to binary.

Reply:
Thanks!